### PR TITLE
Call variation through callback so it's only loaded when needed - in support of  trac 59969

### PR DESCRIPTION
**Pull request by kt-12**
## What?
This PR will help build variations only when we need to load them instead of during resignation at the `init` hook. 

Core PR with detailed performance analysis -> 
https://github.com/WordPress/wordpress-develop/pull/5718

## Why?
Currently, we run a performance-heavy build variation function for some of the blocks in `init` hook, which causes it to load even on the frontend where it is not needed.

## How?
In this PR we register only the callback during block registration and build the variation only when they are needed.

**Note: This PR provides around ~9% improvement to the frontend load time but it has no improvement for the editor site. However, we have found some way by which similar performance benefits could be obtained at the editor side -  https://github.com/WordPress/wordpress-develop/pull/5718#issuecomment-1845361155**

## Testing Instructions
Test it for the TT4 theme. 
1. Open the editor. 
2. Open the browser console and type `wp.blocks.getBlockType('core/post-terms')`
3. Ensure that the block object returned has variation. Do the same for `core/navigation-link`, `core/template-part`

### CODE REVIEW

File: lib/compat/wordpress-6.4/block-hooks.php [hunk 1]
```diff
@@ -79,6 +79,11 @@ function gutenberg_add_hooked_blocks( $settings, $metadata ) {
 
 	$exposed_settings = array_intersect_key( $settings, $fields_to_pick );
 
+	// If the block has a variations callback, call it and add the variations to the block.
```
Comment (joemcgill) [inline lib/compat/wordpress-6.4/block-hooks.php]: I think we will want to add any compat shims to the wordpress-6.5 folder instead of this one.

Comment (kt-12) [reply lib/compat/wordpress-6.4/block-hooks.php]: The way it is loaded, I had to remove the function entirely from 6.4 and add it to 6.5.

File: lib/compat/wordpress-6.4/block-hooks.php [hunk 2]
```diff
@@ -302,6 +302,20 @@ function gutenberg_register_block_hooks_rest_field() {
 	);
 }
 
+/**
```
Comment (spacedmonkey) [inline lib/compat/wordpress-6.4/block-hooks.php]: This code should not live here. Create a new file in the `lib/compat/wordpress-6.5` directory. Maybe blocks.php.

**Review by spacedmonkey — CHANGES_REQUESTED** [review]
One nitpic, otherwise looks good.

File: lib/compat/wordpress-6.4/block-hooks.php (hunk 2 above)
Comment (kt-12) [reply lib/compat/wordpress-6.4/block-hooks.php]: I have moved the code to `lib/compat/wordpress-6.5/blocks.php`

**Review by spacedmonkey — APPROVED** [review]
Great work @kt-12 . 

I can merge this once the automated tests pass.

**Comment by Mamaduka** [comment]
I believe this PR won't need a manual PHP backport.

**Comment by tellthemachines** [comment]
@Mamaduka doesn't the change in `lib/compat/wordpress-6.5/blocks.php` need to be added to core?

**Comment by Mamaduka** [comment]
@tellthemachines, no, it's a shim for a new property to work with older WP versions. It can be removed after the required version is 6.5.

**Comment by getdave** [comment]
✅ I updated the [PHP Sync Tracking Issue](https://github.com/WordPress/gutenberg/issues/57959) to note this PR does not require a backport.

File: phpcs.xml.dist [hunk 3]
```diff
@@ -115,6 +115,8 @@
 				<element value="apply_block_core_search_border_style"/>
 				<element value="apply_block_core_search_border_styles"/>
 				<element value="build_dropdown_script_block_core_categories"/>
+				<element value="build_navigation_link_block_variations"/>
```
Comment (anton-vlasenko) [inline phpcs.xml.dist:118]: @kt-12, I appreciate your efforts with the ignore list. However, these functions weren't part of the initial setup and don't pose any backward compatibility issues if we stick to the naming conventions enforced by the sniff.
For example, they could be renamed to `block_core_template_part_build_area_variations()` and `block_core_post_terms_build_variations()` respectively. Could you please rename them?

Comment (kt-12) [reply phpcs.xml.dist:118]: I added this as I was getting an error from GitHub action. Adding this did resolve the error back then is what I remember, but I am not sure at the moment. Should we remove it?

Comment (Mamaduka) [reply phpcs.xml.dist:118]: Here is the follow-up #58538.

Comment (anton-vlasenko) [reply phpcs.xml.dist:118]: The names I suggested for these functions were not exactly correct, but the idea was to adhere to the coding standard. Thanks for fixing it, @Mamaduka.
I have approved your PR.

**Pull request merged**